File: hyejin/Week_14/Linked List Cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        HashSet<ListNode> history = new HashSet<>();

        if(head == null)
            return false;
        
        ListNode curNode = head;
		
        while(curNode.next != null) {
			ListNode next = curNode.next;

			if(history.contains(next))
				return true;
            else {
				history.add(next);
                curNode = next;                
            }
		}
		return false;   
    }
}

File: sangryeol/Week_14/Linked List Cycle.cpp
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* pCurr = head;
        ListNode* pNext = head;
        while (pNext && pNext->next) {
            pCurr = pCurr->next;
            pNext = pNext->next->next;
            if (pCurr == pNext) return true;
        }
        return false;
    }
};